feat(enhancement): Add support for AVIF and image sequences

tibetiroka opened the pull request:
**Feature**

This PR addresses the feature described in issue #4980. Also related to #10064 and #9880. (Fixes #4980.)

## Summary
Adds support for AVIF/AVIFS files that use the AV1 Image File Format using libavif. These files can store standalone images or image sequences.

Since some of our code assumed three-character file extensions, some marginally-related code had to be refactored. This also enabled using `.jpeg` instead of `.jpg`. (These changes are split off to #10392, but are still necessary here.)

Due to how image buffers are used, mixing regular images with image sequences (or even just using multiple image sequences) would require a serious refactor (mainly because the frame count is not known when we start loading images). Therefore I chose to only support image sequences as standalone; when one of the files for a sprite is an image sequence, all other files are discarded.

## Screenshots
An Albatross loaded from an avif file.
![image](https://github.com/user-attachments/assets/745d463f-8e6a-4ac6-b4ac-25de98551ab9)

## Usage examples
Much better compression for images (and better quality), easier management and even greater savings on animations.

## Testing Done
Some AVIF ship sprites for testing; one is a sequence, the other one is a single image.
[ship.zip](https://github.com/user-attachments/files/16562429/ship.zip)

I tested that images and image sequences are loaded and displayed. I tested that swizzle masks can be calculated. (Please note that the albatross image has some loose pixels; this is due to the crappy online converter used to turn it into avif.)

## Wiki Update
https://github.com/endless-sky/endless-sky-wiki/pull/41

## Performance Impact
AVIF should actually speed things up, especially if we factor in the less disk I/O.

Based on [Google's report](https://storage.googleapis.com/avif-comparison/index.html), AVIF has great performance.
![image](https://github.com/user-attachments/assets/6d7653db-08bb-47a5-8748-06b35aea2029)
![image](https://github.com/user-attachments/assets/ebd8d860-f595-47ee-8abb-289e19114e80)

tibetiroka commented on source/ImageBuffer.cpp:
```diff
@@ -352,6 +369,143 @@ namespace {
+		}
+
+		// Now that we know the buffer's frame count, we can allocate the memory for it.
+		// If this is an image sequence, the preconfigured frame count is wrong.
```
The constructor and Clear, which are used in ImageSet.

tibetiroka commented on source/ImageBuffer.cpp:
```diff
@@ -352,6 +368,144 @@ namespace {
+		if(!decoder)
+		{
+			Logger::LogError("Could not create avif decoder");
+			return 0;
```
Well it's mainly a memory allocation issue, so it might randomly fail and then work again...

tibetiroka replied:
I mean it might fail without any issues related to avif (and that's the only case it can fail, I believe)

tibetiroka commented on source/ImageBuffer.cpp:
```diff
@@ -352,6 +369,143 @@ namespace {
+	int ReadAVIF(const string &path, ImageBuffer &buffer, int frame, bool alphaPreMultiplied)
+	{
+		// Guard dynamically allocated objects to prevent memory leaks.
+		struct DecoderGuard
```
Yeah it's ugly, but at least it's safer than what we do in the other functions.

tibetiroka replied:
Oh, that's smart.

tibetiroka commented on source/ImageBuffer.cpp:
```diff
@@ -151,39 +157,50 @@ void ImageBuffer::ShrinkToHalfSize()
+				Premultiply(*this, i, additive);
 	}
+
 	return true;
```
Actually, this is reporting the old value. Oops.

tibetiroka commented on source/ImageSet.cpp:
```diff
@@ -146,18 +144,24 @@ namespace {
+
+	string ToLowerCase(std::string str)
+	{
+		std::transform(str.begin(), str.end(), str.begin(), [](unsigned char c){ return std::tolower(c); });
```
It's the best I know without pulling in boost.

tibetiroka commented on source/ImageBuffer.cpp:
```diff
@@ -52,8 +54,11 @@ ImageBuffer::~ImageBuffer()
 {
-	delete [] pixels;
-	pixels = nullptr;
+	if(pixels)
```
Oh, that's a nice change from free().

tibetiroka commented on source/ImageBuffer.cpp:
```diff
@@ -352,6 +369,134 @@ namespace {
+	// logic that avoids duplicating the frames.
+	int ReadAVIF(const string &path, ImageBuffer &buffer, int frame, bool alphaPreMultiplied)
+	{
+		unique_ptr<avifDecoder, void(*)(avifDecoder *)> decoder(avifDecoderCreate(), avifDecoderDestroy);
```
Well it compiles and it sounds like it should work...

Koranir commented on source/image/ImageSet.cpp:
```diff
@@ -24,76 +25,74 @@ this program. If not, see <https://www.gnu.org/licenses/>.
+		return (IsBlend(path[pos]) ? pos : base.length());
+	}
+
+	// Determine whether the given path is to an @2x image.
+	bool Is2x(const string &path)
+	{
+		return path.find("@2x") != string::npos;
+	}
+
+	// Determine whether the given path is to a swizzle mask.
+	bool IsSwizzleMask(const string &path)
+	{
+		return path.find("@sw") != string::npos;
+	}
+
+	// Determine whether the given path or name is to a sprite for which a
+	// collision mask ought to be generated.
+	bool IsMasked(const string &path)
+	{
+		if(path.starts_with("ship/"))
+			return true;
+		if(path.starts_with("asteroid/"))
+			return true;
+
+		return false;
 	}
```
Why move these functions?

tibetiroka replied:
They were originally being used in some other function, so I moved them to have them defined first. I should check if it's still necessary.

Koranir commented on source/image/ImageSet.cpp:
```diff
@@ -24,76 +25,74 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 	{
 		// Get the character index where the "name" portion of the path ends.
-		// A path's format is always: <name>(<blend><frame>)(@sw)(@2x).(png|jpg)
+		// A path's format is always: <name>(<blend><frame>)(@sw)(@2x).(<extension>)
```
```suggestion
		// A path's format is always: <name>(<blend><frame>)(@sw)(@2x).<extension>
```

Koranir commented on source/image/ImageSet.cpp:
```diff
@@ -146,18 +145,24 @@ namespace {
 		}
 	}
+
+	string ToLowerCase(std::string str)
+	{
+		std::transform(str.begin(), str.end(), str.begin(), [](unsigned char c){ return std::tolower(c); });
+		return str;
+	}
```
Reimplementing `Format::LowerCase`?

Koranir commented on source/image/ImageSet.cpp:
```diff
@@ -146,18 +145,24 @@ namespace {
 
-	string ext = path.substr(path.length() - 4);
-	return (ext == ".png" || ext == ".jpg" || ext == ".PNG" || ext == ".JPG");
+	string ext = path.substr(path.find_last_of('.') + 1);
```
Just calculate `path.find_last_of('.')` at the beginning and also use it for the early return, it will also return `string::npos` if no matches are found.

Koranir commented on source/image/ImageSet.cpp:
```diff
@@ -232,6 +234,21 @@ void ImageSet::ValidateFrames() noexcept(false)
 
+	// Ensure that image sequences aren't mixed with other images.
+	for(int i = 0; i < 4; i++)
+		for(auto &str : paths[i])
```
```suggestion
		for(const auto &str : paths[i])
```

Koranir reviewed (changes requested):
Looks good.

tibetiroka commented:
In theory, almost all extension-related changes are present there. I'll make sure to apply your review changes there as well.

tibetiroka commented:
I'm marking this as draft for now, as @warp-core requested a bigger refactor of ImageSet for #10392. I'll un-draft this when the necessary changes are merged.

emixa-d commented:
AVIF might be smaller, but now you need another library. Is size(libavif)+size(images in avif)<size(images in png)?  (I'm not counting libpng since it probably is installed anyways (assuming the OS has good package management) -- but maybe the same applies to libavif, I don't know.)

TomGoodIdea commented on docs/readme-cmake.md:
````diff
@@ -59,7 +59,7 @@ In addition to the below dependencies, you will also need CMake 3.16 or newer, h
 
 ```
-g++ cmake ninja-build curl libsdl2-dev libpng-dev libjpeg-dev libgl1-mesa-dev libglew-dev libopenal-dev libmad0-dev uuid-dev
+g++ cmake ninja-build curl libsdl2-dev libpng-dev libjpeg-dev libgl1-mesa-dev libglew-dev libopenal-dev libmad0-dev uuid-dev libavif-dev
````
I'd move it to the rest of image formats libs
```suggestion
g++ cmake ninja-build curl libsdl2-dev libpng-dev libjpeg-dev libavif-dev libgl1-mesa-dev libglew-dev libopenal-dev libmad0-dev uuid-dev
```

TomGoodIdea commented on docs/readme-cmake.md:
````diff
@@ -74,7 +74,7 @@ While sufficient versions of other dependencies are available, Ubuntu 22.04 does
 
 ```
-gcc-c++ cmake ninja-build SDL2-devel libpng-devel libjpeg-turbo-devel mesa-libGL-devel glew-devel openal-soft-devel libmad-devel libuuid-devel
+gcc-c++ cmake ninja-build SDL2-devel libpng-devel libjpeg-turbo-devel mesa-libGL-devel glew-devel openal-soft-devel libmad-devel libuuid-devel libavif-devel
````
```suggestion
gcc-c++ cmake ninja-build SDL2-devel libpng-devel libjpeg-turbo-devel libavif-devel mesa-libGL-devel glew-devel openal-soft-devel libmad-devel libuuid-devel
```

TomGoodIdea commented on docs/readme-developer.md:
```diff
@@ -28,6 +28,7 @@ Use your favorite package manager to install the following (version numbers may
 
 Additionally, if you want to build unit tests:
    - catch2
+   - libavif-dev
```
shouldn't it be in the main list?

TomGoodIdea commented:
conflicts (probably the readme)

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -169,26 +182,34 @@ void ImageBuffer::ShrinkToHalfSize()
-	if(isJPG && !ReadJPG(data.path, *this, frame))
-		return false;
+	int startFrame = frame;
+	int endFrame = startFrame;
```
Why are you adding startFrame here if you're subtracting it at the end anyway?

tibetiroka replied:
It was done that way to make the refactor easier, but isn't necessary anymore. I replaced it with a regular counter.

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,136 @@ namespace {
+		avifImageTiming timing;
+		for(int i = 0; i < decoder->imageCount; ++i)
+		{
+			result = avifDecoderNthImageTiming(decoder.get(), 0, &timing);
```
Isn't this supposed to be the frame index?
```suggestion
			result = avifDecoderNthImageTiming(decoder.get(), i, &timing);
```

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,136 @@ namespace {
+				Logger::LogError("Could not get image timing for '" + path.generic_string() + "': " + avifResultToString(result));
+				return 0;
+			}
+			if(frameTimeUnit == -1 || (frameTimeUnit > timing.duration && timing.duration))
```
```suggestion
			if(frameTimeUnit < 0 || (frameTimeUnit > timing.duration && timing.duration))
```
or use std::optional.

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,136 @@ namespace {
+		size_t bufferFrameCount = 0;
+		for(size_t i = 0; i < static_cast<size_t>(decoder->imageCount); ++i)
+		{
+			result = avifDecoderNthImageTiming(decoder.get(), 0, &timing);
```
As above.
```suggestion
			result = avifDecoderNthImageTiming(decoder.get(), i, &timing);
```

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,136 @@ namespace {
+
+			avifRGBImage image;
+			avifRGBImageSetDefaults(&image, decoder->image);
+			image.depth = 8; // force 8-bit color depth
```
```suggestion
			image.depth = 8; // Force 8-bit color depth.
```

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,136 @@ namespace {
+			}
+
+			// Now copy the image in the buffer to match frame timings.
+			for(size_t i = 1; i < repeats[avifFrameIndex]; ++i)
```
What about frames that exist, but aren't repeated (repeat == 1)?

tibetiroka replied:
They are already loaded into the buffer once; I'm just making extra copies.

TomGoodIdea replied:
ah ok

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,136 @@ namespace {
+			bufferFrame += repeats[avifFrameIndex];
+
+			avifFrameIndex++;
+			avifRGBImageSetDefaults(&image, decoder->image);
```
`image` will be destroyed anyways, no?
```suggestion
```

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,136 @@ namespace {
+			avifRGBImageSetDefaults(&image, decoder->image);
+		}
+
+		if(avifFrameIndex != decoder->imageCount || bufferFrame != bufferFrameCount)
```
`bufferFrameCount` sums `repeats` values, and in this loop we skip only the entries that have `repeats` == 0, so I don't think we need this, but do as you prefer.
```suggestion
		if(avifFrameIndex != decoder->imageCount)
```

tibetiroka replied:
I know, but extra checks can't hurt.

TomGoodIdea commented on source/image/ImageSet.cpp:
```diff
@@ -173,24 +188,41 @@ void ImageSet::Load() noexcept(false)
-	if(makeMasks)
-		masks.resize(frames);
+
+	const auto UpdateFrameCount = [&]() {
```
```suggestion
	const auto UpdateFrameCount = [&]()
	{
```

thewierdnut commented:
I know I'm late to the party, and I'm likely going to get ignored here, but I'm going to add my two cents anyways. 

### I don't think avif will improve loading speed. If you have evidence to the contrary, I would love to see it.
1. avif might be faster compare to other "advanced" graphics format, but it is typically *slower* than the already existing support for png and jpeg. 
2. avif produces smaller files, but disk access is *not* the bottleneck here. Image post-processing is. Note that as part of the loading process, we pre-multiply all of the images, and compute collision masks for all the sprites.
3. Note that the desktop build of endless-mobile *does* have reduced loading times. It does this by premultiplying all images prior to applying etc2 texture compression. This makes the file size much larger than the original png and jpgs, but the files are smaller in memory than the raw decompressed jpg/pngs. The compressed textures can be directly loaded onto the graphics card more quickly, and the collision masks can be computed without needing to decompress them as well. 

### This feature (in my opinion) is not worth the additional maintenance cost.
1. The multi-frame support sounds nice, but we already support that using image sequences.
2. This will add a new external dependency to the code, without solving any real problems or adding any real improvement.
3. In general, adding new libraries or upgrading old ones should only happen if it adds a new feature, fixes a bug, or avoids deprecation. I don't see any of those cases applying here.

tibetiroka commented:
> I don't think avif will improve loading speed. If you have evidence to the contrary, I would love to see it.

> avif might be faster compare to other "advanced" graphics formats, but it is typically slower than the already existing support for png and jpeg.

I did some testing by encoding various animations with avif, and running the game with the `--parse-assets` option. The execution time was measured by `time`; I'm showing the averages and standard deviations of the three runs rounded to three digits. I did three successive runs in each configuration, with a preceding dry run to ensure similar filesystem caching behaviour. The image data was on an M.2 SSD. This should minimize the I/O performance differences. (This is somewhat unfair to avif, as the smaller file sizes wouldn't hurt it that much on slower drives.)

First, I tested the current game images. Then I replaced the asteroid animations with avif. In the third round, I replaced every standalone "animation directory" with avif. In the fourth round, I did all these replacements for high dpi as well.
![image](https://github.com/user-attachments/assets/4aed514c-4ec7-4303-8ec2-c647572ccdab)

As you can see, avif is _not_ slower than png; in fact, it seems to be ever so slightly faster, even when trying to not factor in the file size differences.

> avif produces smaller files, but disk access is not the bottleneck here. Image post-processing is. Note that as part of the loading process, we pre-multiply all of the images, and compute collision masks for all the sprites.

Avif handles alpha premultiplication natively, which lets us skip a processing step later. This is probably good for performance.

> The multi-frame support sounds nice, but we already support that using image sequences.

Like Koranir already said, the current support sucks. You can't easily preview an animation on github, you have to change dozens of files to change an animation. It's not nearly as convenient as an animated format. (Whether that format is avif or something else doesn't really matter.)

> This will add a new external dependency to the code, without solving any real problems or adding any real improvement.

It solves the animation problems outlined above, and greatly reduces file sizes. (In my quick test, the asteroids in the game went from ~15MB to 3MB.)

thewierdnut commented:
> Avif handles alpha premultiplication natively, which lets us skip a processing step later. This is probably good for performance.

Does that mean that your measured numbers above are also skipping the alpha premultiplication in the loading code?

thewierdnut commented:
I would note that switching to pre-multiplied compressed textures reduces the loading time on my PC from 6.6 seconds from 15.1 seconds. This is almost all preprocessing, since I don't think my driver does native etc2 textures. On my phone, which *does* support native etc2 textures, the loading time is about 3 seconds, vs about 30 seconds.

tibetiroka commented:
> Does that mean that your measured numbers above are also skipping the alpha premultiplication in the loading code?

I haven't checked in detail, but avif uses the YUV color space internally, and converts it to RGB while decoding. This conversion is different with premultiplied alpha, but I think this would only affect images that already have their alpha values premultiplied.

thewierdnut commented:
> > Does that mean that your measured numbers above are also skipping the alpha premultiplication in the loading code?
> 
> I haven't checked in detail, but avif uses the YUV color space internally, and converts it to RGB while decoding. This conversion is different with premultiplied alpha, but I think this would only affect images that already have their alpha values premultiplied.

Ah, I'm not all the interested in what avif is doing internally, I just wanted to check if you were still doing pre-multiplication during the post-processing stage after the image is loaded. It sounds like you are.

warp-core commented on source/image/ImageSet.cpp:
```diff
@@ -146,6 +146,21 @@ void ImageSet::ValidateFrames() noexcept(false)
 	framePaths[3].clear();
 
+	// Ensure that image sequences aren't mixed with other images.
+	for(int i = 0; i < 4; i++)
```
Style
```suggestion
	for(int i = 0; i < 4; ++i)
```

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,135 @@ namespace {
+			throw runtime_error(message);
+		}
+		if(static_cast<unsigned>(buffer.Width()) != decoder->image->width
+				|| static_cast<unsigned>(buffer.Height()) != decoder->image->height)
```
```suggestion
			|| static_cast<unsigned>(buffer.Height()) != decoder->image->height)
```

TomGoodIdea commented on source/image/ImageBuffer.cpp:
```diff
@@ -359,6 +379,135 @@ namespace {
+			}
+			bufferFrame += repeats[avifFrameIndex];
+
+			avifFrameIndex++;
```
```suggestion
			++avifFrameIndex;
```
